File: lab1/blur_float.py
"""
4190.308 Computer Architecture                                                          Spring 2023

Image blurring (float)

This module implements a function that blurs an image with a 3x3 filter (floating point version)

@author:
    Your Name <your@email.com>

@changes:
    2023/MM/DD Your Name Comment

"""



def blur(image, height, width, channels, kernel_size=5):
    """
    Blurs an image with a kernel and returns the blurred image.

    Args:
        image:        image data (multi-level list)
        height:       image height
        width:        image width
        channels:     number of channels (BGR or BGRA)
        kernel_size:  size of blurring kernel

    Returns:
        A tuple containing the following elements:
        - blurred:    blurred image data
        - bheight:    blurred image height
        - bwidth:     blurred image width
        - bchannels:  blurred image channels

    """

    # TODO
    # Your work goes here

    # For now, we simply copy the input parameters into the output parameters.
    # Fix/adjust once you have implemented your solution.
    bheight = height + 1 - kernel_size
    bwidth = width + 1 - kernel_size
    bchannels = channels
    blurred = []

    weight = 1/ (kernel_size**2)

    for h in range(bheight):
        h_temp = []
        for w in range(bwidth):
            w_temp = []
            for c in range(bchannels):
                temp = 0
                for x in range(kernel_size):
                    for y in range(kernel_size):
                        temp += image[h+x][w+y][c]/255
                temp = temp * weight
                w_temp.append(int(temp*255))
            h_temp.append(w_temp)
        blurred.append(h_temp)

    return blurred, bheight, bwidth, bchannels

File: lab2/blur_float.c
//-------------------------------------------------------------------------------------------------
// 4190.308 Computer Architecture                                                       Spring 2023
//
/// @file
/// @brief Image blurring (float)
///        This module implements a function that blurs an image with a filter (floating-point 
///        version). Note that CPython uses the 'double' data type for floating point numbers, 
///        so one may observe small differences in the output if the 'float' data type is used.
///
/// @author Your Name <your@email.com>
///
/// @section changelog Change Log
/// 2023/MM/DD Your Name Comment
///
//-------------------------------------------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include "blur.h"


struct Image blur_float(struct Image image, int kernel_size)
{
  // TODO
  // Your work goes here
  struct Image bimage;
  int h, w, c, x, y;
  float temp;
  bimage.height = image.height + 1 - kernel_size;
  bimage.width = image.width + 1 - kernel_size;
  bimage.channels = image.channels;

  float weight = (float)1.0/(float)(kernel_size*kernel_size);

  bimage.data = (uint8*)malloc(sizeof(uint8*) * bimage.height * bimage.width * bimage.channels);
  for (h=0; h < bimage.height; h++){
    for (w=0; w < bimage.width; w++){
      for (c=0; c < bimage.channels; c++){
        temp = 0;
        for (x=0; x<kernel_size; x++){
          for (y=0; y<kernel_size; y++){
            temp += (float) PIXEL(image, h+x, w+y, c) / (float) 255.0;
          }
        }
        temp = temp * weight;
        temp = temp * 255.0;
        bimage.data[INDEX(bimage, h, w, c)] = (uint8)(temp);
      }
    }
  }
  // For now, we simply return the input image.
  // Fix/adjust once you have implemented your solution.
  return bimage;
}
